### ISSUE TITLE
feat(TypeScript): enable 'noImplicitThis' options

### ISSUE THREAD

**Issue by saneyuki** [ReactiveX/rxjs#1971]
- This intent to make `this` type explicitly.
  - This would make our codes more type safe with a compiler enforcement.

**Comment by saneyuki** [comment]
Duplicated #1837

**Comment by saneyuki** [comment]
Relate? https://github.com/ReactiveX/rxjs/pull/1978

**Comment by kwonoj** [comment]
I'm bit down for this. I've enabled this option for few codebases, and found it's sometimes not easy to infer correct type of `this` and eventually had to go down with explicitly `any` casting still. Problem of `this` is it's reserved keyword, you can't just cast simply like `this as any`. For example, this is actual erroring code in Rx codebase after enabling this option - which just need additional boilerplate to casting.

```js
//'this' implicitly has type 'any' because it does not have a type annotation.
state.context.closeAction = (<any>this).schedule(state, state.bufferTimeSpan); 
````

**Comment by saneyuki** [comment]
@kwonoj 

We can specify `this` as 1st argument of a signature like `(this: A) => void` or the following:

```ts
function dispatchBufferTimeSpanOnly<T>(this: Action<T>, state: any) {
    ...
    // this `this` is `Action<T>`
    state.context.closeAction = this.schedule(state, state.bufferTimeSpan);
}
```

If we land https://github.com/ReactiveX/rxjs/pull/2078, we can checks validness of it statically.

**Comment by kwonoj** [comment]
> We can specify this as 1st argument of a signature like (this: A) => void or the ...

Yes I am aware of those. Question in here is is it worth to do it? as our codebase shows, when we use `this`, we are aware its usage and explicitly cast to `any` anyway. So enabling this option only requires additional boilerplate to cast into any since most of usage those not asks explicit type inference. Above snippet can be addressed via #2078  as you mentioned though, it is about general of accessing `this` in certain context.

**Comment by saneyuki** [comment]
@kwonoj 



- I think this would  clarify `this` type for callbacks (e.g. `Observable<T>`'s constructor, `Scheduler.schedule()`). The signature express very strongly what this `this`. 
   - This increase our code's readability. It is nice for us and users.
- And, there is no (or less) cons as my feeling by that I'm trying to enable this option in my private branch.
   - In almost case, we need not add additional boilerplate. We need specify `this` type for only a function which uses implicit this type. If a function does not use `this`, then we don't have to specify `this` type.
   -  If we need to cast `this` to some derived type of it, we need to use temp variables like these, but I seem there are rare cases actually, and we used its approach in the current code. It would not be a serious constraints problem for us. 
    - https://github.com/ReactiveX/rxjs/blob/master/src/observable/BoundNodeCallbackObservable.ts#L128
    - https://github.com/ReactiveX/rxjs/blob/master/src/testing/ColdObservable.ts#L24
